Source/Core/Common/UPnP.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#ifdef USE_UPNP
+
+#include "Common/UPnP.h"
+
+#include "Common/Logging/Log.h"
+
+#include <array>
+#include <cstdlib>
+#include <cstring>
+#include <miniupnpc.h>
+#include <miniwget.h>
+#include <string>
+#include <thread>
+#include <upnpcommands.h>
+#include <vector>
+
+static struct UPNPUrls s_urls;
+static struct IGDdatas s_data;
+static std::array<char, 20> s_our_ip;
+static u16 s_mapped = 0;
+static std::thread s_thread;
+
+// called from ---UPnP--- thread
+// discovers the IGD
+static bool InitUPnP()
+{
+  static bool s_inited = false;
+  static bool s_error = false;
+
+  // Don't init if already inited
+  if (s_inited)
+    return true;
+
+  // Don't init if it failed before
+  if (s_error)
+    return false;
+
+  std::memset(&s_urls, 0, sizeof(UPNPUrls));
```

Comment:
Is it possible to remove these two memset calls and instead value initialise s_urls and s_data directly?